doc/source/workflows/metadata.rst

```diff
@@ -0,0 +1,208 @@
+
+**Note: user-defined metadata must be a python dictionary with values that are
+JSON serializable.**
+
+Metadata in Virtual Actor
+-------------------------
+Virtual Actors also support metadata ingestion and retrieval. For example:
+
+.. code-block:: python
+
+    @workflow.virtual_actor
+    class Actor:
+        def __init__(self, v):
+            self.v = v
+
+        def add_v(self, v):
+            time.sleep(1)
+            self.v += v
+            return self.v
+
+    actor = Actor.get_or_create("vid", 0)
+
+    actor.add_v.options(name="add", metadata={"k1": "v1"}).run(10)
+    actor.add_v.options(name="add", metadata={"k2": "v2"}).run(10)
+    actor.add_v.options(name="add", metadata={"k3": "v3"}).run(10)
+
+    assert workflow.get_metadata("vid", "add")["user_metadata"] == {"k1": "v1"}
+    assert workflow.get_metadata("vid", "add_1")["user_metadata"] == {"k2": "v2"}
+    assert workflow.get_metadata("vid", "add_2")["user_metadata"] == {"k3": "v3"}
+    assert workflow.get_metadata("vid", "add")["stats"]["end_time"] >= workflow.get_metadata("vid", "add")["stats"]["start_time"] + 1
+    assert workflow.get_metadata("vid", "add_1")["stats"]["end_time"] >= workflow.get_metadata("vid", "add_1")["stats"]["start_time"] + 1
+    assert workflow.get_metadata("vid", "add_2")["stats"]["end_time"] >= workflow.get_metadata("vid", "add_2")["stats"]["start_time"] + 1
+
+Notice that if there are multiple steps with the same name, a suffix
+with a counter _n will be added automatically.
+
+And you can also do this in a nested way:
+
+.. code-block:: python
+
+    @workflow.virtual_actor
+    class Counter:
+        def __init__(self):
+            self.n = 0
+
+        def incr(self, n):
+            self.n += 1
+            if n - 1 > 0:
+                return self.incr.options(
+                    name="incr", metadata={
+                        "current_n": self.n
+                    }).step(n - 1)
+            else:
+                return self.n
+
+    counter = Counter.get_or_create("counter")
+    counter.incr.options(name="incr", metadata={"outer_k": "outer_v"}).run(5)
+
+    assert workflow.get_metadata("counter", "incr")["user_metadata"] == {"outer_k": "outer_v"}
+    assert workflow.get_metadata("counter", "incr_1")["user_metadata"] == {"current_n": 1}
+    assert workflow.get_metadata("counter", "incr_2")["user_metadata"] == {"current_n": 2}
+    assert workflow.get_metadata("counter", "incr_3")["user_metadata"] == {"current_n": 3}
+    assert workflow.get_metadata("counter", "incr_4")["user_metadata"] == {"current_n": 4}
+
+Available Metrics
+-----------------
+**workflow level**
+
+- status: workflow states, can be one of RUNNING, FAILED, RESUMABLE, CANCELED, or SUCCESSFUL.
+- user_metadata: a python dictionary of user input metadata that is passed in ``workflow.run()``.
+- stats: workflow running stats, including workflow start time and end time.
+
+**step level**
+
+- name:
+- workflow:
+- workflow_refs:
+- step_options:
```

Comment:
```
- name: The name of the step. It's either provided by the user via ``step.options()`` or generated by the system.
```

We can remove workflow/workflow_refs for now.

Reply:
@iycheng the "name" reminds me of something that I intended to open an issue last month - 
https://github.com/lchu-ibm/ray/blob/metadata-doc/python/ray/workflow/common.py#L192
I am wondering if here we should do a 
`"name": self.name or get_module(f) + "." + get_qualname(f),` as we have added the support of named step. I feel this is something should be changed.  Because the inputs.json is a dump version from this `to_metadata` method and we should use the name of the step when a user gives one.